Hanging attributes

AndrewRadev opened the pull request:
This PR implements a "hanging" attribute style. If the first line of the component has at least one attribute set on that line, the rest align according to that. So, it looks like this:

``` handlebars
{{foo-bar one="two"
          two="three"
          four="five"}}
```

If the first line only has the component name, and not an attribute, it indents like it did before.

Unfortunately, I branched this PR off from my own master, which includes PRs #53 and #52. Depending on what you decide, I can rebase this on top of your master, but for now, I'll just leave it here for consideration as-is.

nelstrom commented:
I'd love to see this merged. In the meantime, I've switched to using `AndrewRadev/vim-mustache-handlebars`.

shadoath commented:
Bump. Is this repo being actively maintained?

jsit commented:
@AndrewRadev Can you please resolve these conflicts?

AndrewRadev commented:
I've rebased the branch on top of the current master (or rather, cherry-picked the changes without the variable renames), and I've fixed a bug I'd introduced. I'd recommend testing it out before merging, though, indentation is tricky.

AndrewRadev commented:
I've merged the most recent master, and tested it out a bit, and it seems to work well now after #93. If you have time to test it out and find any issues to fix, let me know.

jsit commented:
Thanks @AndrewRadev. Can you please add some test cases to example.mustache?

AndrewRadev commented:
@jsit I've pushed an example of component with hanging attributes, and I've also applied a fix for that file -- it had an unclosed `<a>` tag.

That said, I see there's a new example there of an angle-brackets component that doesn't seem to indent correctly:

``` mustache
<AngleBracketComponent
@mustache={{@bla}}
@number=123
data-test-id="test"
class="my-class" as |test|>
Some Text
<div>{{test.someValue}}</div>
<test.component teste="asdfads"></test.component>
</AngleBraketComponent>
```

But actually, even when I change to master, it still doesn't indent correctly, so that doesn't seem to be a regression from this PR. It might be something that's just not handled on my config and expects maybe HTML support to deal with it.

Well, either way, it seems like an unrelated issue.

jsit commented:
@AndrewRadev Thanks; I've already got an issue for the angle bracket indent issue, #92

Will have a look at your hanging example now, thank you!

jsit commented:
@AndrewRadev Also curious to get your opinion on #98

jsit commented on indent/handlebars.vim:
```diff
@@ -79,16 +79,58 @@ function! GetHandlebarsIndent(...)
 
-  " indent after block {{#block
-  if prevLine =~# '\v\s*\{\{[#^].*\s*'
+  " " check for a hanging attribute
```
```suggestion
  " check for a hanging attribute
```

jsit commented on indent/handlebars.vim:
```diff
@@ -79,16 +79,58 @@ function! GetHandlebarsIndent(...)
+    endif
+  endif
+
+  let componentClosingPattern = '\v^\s*\{\{\/\S*\}\}\s*'
```
Can we move this definition closer to the beginning of the script, and scope it with `s:`?

jsit commented on indent/handlebars.vim:
```diff
@@ -79,16 +79,58 @@ function! GetHandlebarsIndent(...)
+
+  " indent after block: {{#block, {{^block
+  if prevLine =~# '\v\s*\{\{[#^].*\s*' &&
+        \ prevLine !~# '{{[#^]\(.\{}\)\s.\{}}}.*{{\/\1}}'
```
Is this additional check separate from the purpose of this PR? This looks unrelated to hanging attributes to me but I might be misreading it.

AndrewRadev replied:
I think it's not related. I guess I've added it to avoid increasing indent for block components on the same line, but the lines below seem to take care of that.

I removed the check locally, and it seems like it doesn't affect anything. I'll commit a removal of it, and if you see any regressions, we can put it back, I guess.

jsit replied:
Thanks -- if it's a useful change, I'd love to see it in a separate PR

AndrewRadev commented:
I think I've addressed your comments -- let me know what you think.

> @AndrewRadev Also curious to get your opinion on #98

Yes, that seems like a very sensible change. I remember being confused by the explicit `runtime`s of syntax and indent scripts, it would be much easier to understand to just have separate `handlebars` and `mustache` files where one includes the other.

jsit commented on indent/handlebars.vim:
```diff
@@ -79,16 +81,55 @@ function! GetHandlebarsIndent(...)
+    endif
+  endif
+
+  " check for a closing }}, indent according to the opening one
```
Is this part of hanging indent, too? Or is this a separate improvement? I just want to keep this PR as clean as possible.

jsit replied:
I see now that this is necessary for the hanging indent. However, it looks like it duplicates some other functionality.

I notice that on this line:

https://github.com/mustache/vim-mustache-handlebars/pull/55/files#diff-eeee5d46b419bb17834d0b1a86a43f8fR114

It's returning `indent(line) + sw`, which is also happening later:

```vim
  " indent after block: {{#block, {{^block
  if prevLine =~# '\v\s*\{\{[#^].*\s*'
    let ind = ind + sw
  endif
```

Would it be possible to restrict this section to *just* the "hanging indent"-related conditionals?

jsit commented on indent/handlebars.vim:
```diff
@@ -79,16 +81,55 @@ function! GetHandlebarsIndent(...)
+        \ search('}}$', 'Wb')
+    let [line, col] = searchpairpos('{{', '', '}}', 'Wb')
+    if line > 0
+      if strpart(getline(line), col - 1, 3) =~ '{{[#^]'
+        " then it's a block component, indent a shiftwidth more
+        return indent(line) + sw
+      else
+        return indent(line)
+      endif
```
```suggestion
      if strpart(getline(line), col - 1, 3) !=~ '{{[#^]'
        return indent(line)
      endif
```

We could almost do this, see what I mean? the `indent(line) + sw` for "normal" indents is happening later anyway, so it would be nice to keep that there.

jsit commented:
An issue here if I'm *typing out* a mustache element, rather than indenting one with `=`, it stops correctly doing the hanging indent after the second line:

<img width="294" alt="Screen Shot 2020-08-09 at 5 26 13 PM" src="https://user-images.githubusercontent.com/643417/89742164-82d87f80-da65-11ea-9157-d07f27726ad0.png">
